src/limiter/positivity_preserving_limiter.cpp

```diff
@@ -172,8 +187,160 @@ void PositivityPreservingLimiter<dim, nstate, real>::write_limited_solution(
                 std::abort();
             }
+            // Verify that the solution values haven't been changed to NaN as a result of all quad pts in a cell having negative density 
+            // (all quad pts having negative density would result in the local maximum convective eigenvalue being zero leading to division by zero)
+            if (isnan(solution[current_dofs_indices[idof]])) {
+                std::cout << "Error: Solution is NaN - Aborting... " << std::endl << solution[current_dofs_indices[idof]] << std::endl << std::flush;
+                std::abort();
+            }
+        }
+    }
+}
+
+template <int dim, int nstate, typename real>
+std::array<real, nstate> PositivityPreservingLimiter<dim, nstate, real>::get_soln_cell_avg_PPL(
+    const std::array<std::array<std::vector<real>, nstate>, dim>&        soln_at_q,
+    const unsigned int                                                   n_quad_pts,
+    const std::vector<real>&                                             quad_weights_GLL,
+    const std::vector<real>&                                             quad_weights_GL,
+    double&                                                              dt)
+{
+    std::array<real, nstate> soln_cell_avg;
+
+    // Obtain solution cell average
+    if (dim == 1) {
+        soln_cell_avg = get_soln_cell_avg(soln_at_q[0], n_quad_pts, quad_weights_GLL);
+    } else if (dim > 1) {
+        std::array<std::array<real, nstate>,dim> soln_cell_avg_dim;
+
+        for(unsigned int idim = 0; idim < dim; ++idim) {
+            for(unsigned int istate = 0; istate < nstate; ++istate) {
+                soln_cell_avg_dim[idim][istate] = 0;
+            }
+        }
+
+        if (dim == 2) {
+            for(unsigned int istate = 0; istate < nstate; ++istate) {
+                unsigned int quad_pt = 0;
+                for(unsigned int iquad=0; iquad<quad_weights_GLL.size(); ++iquad) {
+                    for(unsigned int jquad=0; jquad<quad_weights_GL.size(); ++jquad) {
+                        soln_cell_avg_dim[0][istate] += quad_weights_GLL[iquad]*quad_weights_GL[jquad]*soln_at_q[0][istate][quad_pt];
+                            quad_pt++;
+                    }
+                }
+            }
+
+            for(unsigned int istate = 0; istate < nstate; ++istate) {
+                unsigned int quad_pt = 0;
+                for(unsigned int iquad=0; iquad<quad_weights_GL.size(); ++iquad) {
+                    for(unsigned int jquad=0; jquad<quad_weights_GLL.size(); ++jquad) {
+                        soln_cell_avg_dim[1][istate] += quad_weights_GL[iquad]*quad_weights_GLL[jquad]*soln_at_q[1][istate][quad_pt];
+                            quad_pt++;
+                    }
+                }
+            }
+        }
+
+        if (dim == 3) {
+            for(unsigned int istate = 0; istate < nstate; ++istate) {
```

Comment:
For the 3D case do we need to consider six options or does their paper just limit it to GLL x GL x GL? If they did not it would be interesting to see how the analysis/math turns out.